snipssonos/helpers/spotify_client.py

```diff
@@ -0,0 +1,207 @@
+
+    def _get_authorization_headers_from_client_credentials(self):
+        base64_encoded_credentials = self._get_base_64_encoded_credentials()
+        auth_headers = {'Authorization': 'Basic {}'.format(base64_encoded_credentials)}
+        return auth_headers
+
+    def _get_authorization_headers_from_access_token(self):
+        if not len(self.access_token):
+            raise MusicSearchCredentialsError
+        else:
+            auth_headers = {'Authorization': 'Bearer {}'.format(self.access_token)}
+            return auth_headers
+
+    def retrieve_access_token(self):
+        try:
+
+            headers = self._get_authorization_headers_from_client_credentials()
+
+            response = requests.post(
+                self.AUTH_SERVICE_ENDPOINT,
+                headers=headers,
+                data={'grant_type': 'client_credentials'}
+            )
+
+            # TODO figure what is going with this call
+            # data = {
+            #     'grant_type': 'refresh_token',
+            #     'refresh_token': self.refresh_token
+            # }
+            access_token = self._extract_access_token(response)
+            return access_token
+
+        except requests.exceptions.ConnectionError as e:
+            raise MusicSearchProviderConnectionError
+
+    def authenticate(self):
+        if self.access_token is None:
+            self.access_token = self.retrieve_access_token()
+
+    def execute_query(self, query=None):
+        try:
+            if query is not None:
+                query = query.to_dict()
+            self.authenticate()
+            headers = self._get_authorization_headers_from_access_token()
+            response = requests.get(
+                self.endpoint,
+                params=query,
+                headers=headers)
+            if response.ok:
+                return response.text
+            else:
+                raise MusicSearchProviderConnectionError
+        except requests.exceptions.ConnectionError as e:
+            raise MusicSearchProviderConnectionError
+
+    def set_search_endpoint(self):
+            self.endpoint = self.SEARCH_ENDPOINT
+
+    def set_user_endpoint(self, ):
+        self.endpoint = self.USER_ENDPOINT
+        return self
+
+    def set_top_tracks_endpoint(self):
+        self.is_user_endpoint_set()
+        self.endpoint = "{}/top/{}".format(self.endpoint, "tracks")
+
+    def set_top_artist_endpoint(self):
+        self.is_user_endpoint_set()
+        self.endpoint = "{}/top/{}".format(self.endpoint, "artists")
+
+    def set_playlist_endpoint(self):
+        self.is_user_endpoint_set()
+        self.endpoint = "{}/{}".format(self.endpoint, "playlists")
+
+    def is_user_endpoint_set(self):
+        if self.endpoint != self.USER_ENDPOINT:
+            raise SpotifyClientWrongEndpoint
+
+
+class SpotifyAPISearchQueryBuilder(object):
+    TIME_RANGES = ["long_term", "medium_term", "short_term"]
+
+    def __init__(self, is_request_user_data_query=False):
```

Comment:
I would not pass is_request_user_data_query as a parameter, i would use a specific method as it's a builder.

Reply:
You mean within SpotifyAPISearchQueryBuilder, create a method that enables this option when called?

Reply:
Previous comment will resolve with new implementation of query builder which deals witht he endpoint where to make the request as well.